pkg/querier/queryrange/shard_resolver.go

```diff
@@ -0,0 +1,153 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	math "math"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/concurrency"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/logql/syntax"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/config"
+	"github.com/grafana/loki/pkg/storage/stores/index/stats"
+	"github.com/grafana/loki/pkg/storage/stores/shipper/indexgateway/indexgatewaypb"
+	"github.com/grafana/loki/pkg/util/spanlogger"
+)
+
+func shardResolverForConf(
+	ctx context.Context,
+	conf config.PeriodConfig,
+	defaultLookback time.Duration,
+	logger log.Logger,
+	maxParallelism int,
+	r queryrangebase.Request,
+	handler queryrangebase.Handler,
+) (logql.ShardResolver, bool) {
+	if conf.IndexType == config.TSDBType {
+		return &dynamicShardResolver{
+			ctx:             ctx,
+			logger:          logger,
+			handler:         handler,
+			from:            model.Time(r.GetStart()),
+			through:         model.Time(r.GetEnd()),
+			maxParallelism:  maxParallelism,
+			defaultLookback: defaultLookback,
+		}, true
+	}
+	if conf.RowShards < 2 {
+		return nil, false
+	}
+	return logql.ConstantShards(conf.RowShards), true
+}
+
+type dynamicShardResolver struct {
+	ctx     context.Context
+	handler queryrangebase.Handler
+	logger  log.Logger
+
+	from, through   model.Time
+	maxParallelism  int
+	defaultLookback time.Duration
+}
+
+// from, through, max concurrency to run
+func (r *dynamicShardResolver) Shards(e syntax.Expr) (int, error) {
+	sp, _ := spanlogger.NewWithLogger(r.ctx, r.logger, "dynamicShardResolver.Shards")
+	defer sp.Finish()
+	// We try to shard subtrees in the AST independently if possible, although
+	// nested binary expressions can make this difficult. In this case,
+	// we query the index stats for all matcher groups then sum the results.
+	grps := syntax.MatcherGroups(e)
```

Comment:
I may jump the gun here but what do you think about introducing a query plan that is generated from the AST in the future? This plan could already express if expressions are shardable or not.

Reply:
I'm sure there will be a big query planning rewrite in Loki's future, but let's leave that for a separate task :)

Reply:
Of course. I just wanted to verify that this is out there on our road :slightly_smiling_face: